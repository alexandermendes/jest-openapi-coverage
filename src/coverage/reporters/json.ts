import appRoot from 'app-root-path';
import fse from 'fs-extra';
import path from 'path';
import { JestOpenApiCoverageConfig } from '../../config/openapi';
import { CoverageResult } from '../results';
import { logger } from '../../logger';

export const printJson = (
  results: CoverageResult[],
  config: JestOpenApiCoverageConfig,
) => {
  const { outputFile } = config;

  if (!outputFile) {
    logger.log(JSON.stringify(results));

    return;
  }

  const outputPath = path.isAbsolute(outputFile)
    ? outputFile
    : path.resolve(appRoot.path, outputFile);

  fse.ensureDirSync(path.dirname(outputPath));
  fse.writeJSONSync(outputPath, results);
};
