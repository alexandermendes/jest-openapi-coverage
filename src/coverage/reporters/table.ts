import chalk from 'chalk';
import { Table } from 'console-table-printer';
import { CoverageResult } from '../results';
import { percentageToString } from '../format';

type TableColumnName =
  'method' |
  'path' |
  'queries' |
  'uncoveredQueries';

type TableColumn = {
  name: TableColumnName;
  alignment: 'left' | 'center' | 'right';
  title: string;
}

type TableRow = { [key in TableColumnName]: string }

const MAX_WIDTH = 80;

const colour = (
  text: string,
  conditon: boolean | number,
) => {
  if (typeof conditon === 'boolean') {
    return conditon ? chalk.bold.greenBright(text) : chalk.bold.redBright(text);
  }

  if (conditon === 0) {
    return chalk.bold.redBright(text);
  }

  if (conditon < 100) {
    return chalk.bold.yellowBright(text);
  }

  return chalk.bold.greenBright(text);
};

const addTableRow = (table: Table, result: CoverageResult) => {
  const uncoveredQueries = result
    .queryParams
    .filter(({ covered }) => !covered);

  const row: TableRow = {
    method: colour(result.method.toUpperCase(), result.covered),
    path: colour(result.path, result.covered),
    queries: colour(
      percentageToString(result.percentageOfQueriesCovered),
      result.percentageOfQueriesCovered,
    ),
    uncoveredQueries: colour(`${uncoveredQueries
      .map(({ name }) => name)
      .reduce((acc, name) => {
        if (acc.length < MAX_WIDTH) {
          return [acc, name].filter((str) => str).join(', ');
        }

        if (!acc.endsWith('...')) {
          return `${acc}...`;
        }

        return acc;
      }, '')}`, result.percentageOfQueriesCovered),
  };

  table.addRow(row);
};

export const printTable = (results: CoverageResult[]) => {
  const columns: TableColumn[] = [
    { name: 'method', alignment: 'left', title: 'Method' },
    { name: 'path', alignment: 'left', title: 'Path' },
    { name: 'queries', alignment: 'right', title: '% Queries' },
    { name: 'uncoveredQueries', alignment: 'left', title: 'Uncovered queries' },
  ];

  const table = new Table({ columns });

  results.forEach((result) => addTableRow(table, result));
  table.printTable();
};
