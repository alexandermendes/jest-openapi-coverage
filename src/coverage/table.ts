import { Table } from 'console-table-printer';
import { CoverageResult } from './results';

enum RowColours {
  red = 'red',
  yellow = 'yellow',
  green = 'green',
}

type TableColumnName =
  'method' |
  'endpoint' |
  'queries' |
  'uncoveredQueries';

type TableColumn = {
  name: TableColumnName;
  alignment: 'left' | 'center' | 'right';
  title: string;
}

type TableRow = { [key in TableColumnName]: string }

type TableOptions = { color: RowColours }

const MAX_WIDTH = 80;

const getRowColour = (
  result: CoverageResult,
  percentageQueriesCovered: number,
) => {
  if (!result.covered) {
    return RowColours.red;
  }

  if (percentageQueriesCovered < 100) {
    return RowColours.yellow;
  }

  return RowColours.green;
};

const addTableRow = (table: Table, result: CoverageResult) => {
  const uncoveredQueries = result
    .queryParams
    .filter(({ covered }) => !covered);

  const row: TableRow = {
    method: result.method.toUpperCase(),
    endpoint: result.path,
    queries: result
      .percentageOfQueriesCovered
      .toFixed(2)
      .replace(/0?0$/, '')
      .replace(/\.$/, ''),
    uncoveredQueries: `${uncoveredQueries
      .map(({ name }) => name)
      .reduce((acc, name) => {
        if (acc.length < MAX_WIDTH) {
          return [acc, name].filter((str) => str).join(', ');
        }

        if (!acc.endsWith('...')) {
          return `${acc}...`;
        }

        return acc;
      }, '')}`,
  };

  const options: TableOptions = {
    color: getRowColour(result, result.percentageOfQueriesCovered),
  };

  table.addRow(row, options);
};

export const printTable = (results: CoverageResult[]) => {
  const columns: TableColumn[] = [
    { name: 'method', alignment: 'left', title: 'Method' },
    { name: 'endpoint', alignment: 'left', title: 'Endpoint' },
    { name: 'queries', alignment: 'right', title: '% Queries' },
    { name: 'uncoveredQueries', alignment: 'left', title: 'Uncovered queries' },
  ];

  const table = new Table({ columns });

  results.forEach((result) => addTableRow(table, result));
  table.printTable();
};
