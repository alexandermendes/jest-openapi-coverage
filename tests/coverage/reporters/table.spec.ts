import { Table } from 'console-table-printer';
import { printTable } from '../../../src/coverage/reporters/table';

jest.mock('console-table-printer');
jest.mock('chalk', () => ({
  bold: {
    redBright: (str: string) => `red(${str})`,
    yellowBright: (str: string) => `yellow(${str})`,
    greenBright: (str: string) => `green(${str})`,
  },
}));

describe('Coverage: Table Reporter', () => {
  it('sets up and prints the table', () => {
    printTable([
      {
        method: 'get',
        path: '/articles',
        queryParams: [],
        percentageOfQueriesCovered: 100,
        covered: true,
      },
    ]);

    const mockTables = (Table as jest.Mock).mock.instances;
    const [mockTableCalls] = (Table as jest.Mock).mock.calls;

    expect(mockTables).toHaveLength(1);
    expect(mockTableCalls).toEqual([
      {
        columns: [
          {
            name: 'method',
            alignment: 'left',
            title: 'Method',
          },
          {
            name: 'endpoint',
            alignment: 'left',
            title: 'Endpoint',
          },
          {
            name: 'queries',
            alignment: 'right',
            title: '% Queries',
          },
          {
            name: 'uncoveredQueries',
            alignment: 'left',
            title: 'Uncovered queries',
          },
        ],
      },
    ]);

    expect(mockTables[0].printTable).toHaveBeenCalledTimes(1);
  });

  it.each`
    covered  | percentage | primaryColor   | secondaryColor
    ${true}  | ${100}     | ${'green'}     | ${'green'}
    ${true}  | ${50}      | ${'green'}     | ${'yellow'}
    ${false} | ${100}     | ${'red'}       | ${'green'}
    ${false} | ${0}       | ${'red'}       | ${'red'}
    ${true}  | ${0}       | ${'green'}     | ${'red'}
  `(
    'adds a $color table row for covered=$covered and percentage=$percentage',
    ({
      covered,
      percentage,
      primaryColor,
      secondaryColor,
    }) => {
      printTable([
        {
          method: 'get',
          path: '/articles',
          queryParams: [],
          percentageOfQueriesCovered: percentage,
          covered,
        },
      ]);

      const [mockTable] = (Table as jest.Mock).mock.instances;

      expect(mockTable.addRow).toHaveBeenCalledTimes(1);
      expect(mockTable.addRow).toHaveBeenCalledWith(
        {
          endpoint: `${primaryColor}(/articles)`,
          method: `${primaryColor}(GET)`,
          queries: `${secondaryColor}(${percentage})`,
          uncoveredQueries: `${secondaryColor}()`,
        },
      );
    },
  );

  it('prints a row for partially covered queries', () => {
    printTable([
      {
        method: 'get',
        path: '/articles',
        queryParams: [
          { name: 'one', covered: false },
          { name: 'two', covered: false },
          { name: 'three', covered: true },
        ],
        percentageOfQueriesCovered: 66.6666667,
        covered: true,
      },
    ]);

    const [mockTable] = (Table as jest.Mock).mock.instances;

    expect(mockTable.addRow).toHaveBeenCalledTimes(1);
    expect(mockTable.addRow).toHaveBeenCalledWith(
      {
        endpoint: 'green(/articles)',
        method: 'green(GET)',
        queries: 'yellow(66.67)',
        uncoveredQueries: 'yellow(one, two)',
      },
    );
  });

  it('concatenates many uncovered queries', () => {
    printTable([
      {
        method: 'get',
        path: '/articles',
        queryParams: Array(20).fill(null).map((_, i) => ({
          name: `param${i}`,
          covered: false,
        })),
        percentageOfQueriesCovered: 66.6666667,
        covered: true,
      },
    ]);

    const [mockTable] = (Table as jest.Mock).mock.instances;

    expect(mockTable.addRow).toHaveBeenCalledTimes(1);
    expect(mockTable.addRow.mock.calls[0][0].uncoveredQueries).toMatchSnapshot();
  });
});
