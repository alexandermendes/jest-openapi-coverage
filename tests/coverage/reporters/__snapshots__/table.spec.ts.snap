// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coverage: Table Reporter concatenates many uncovered queries 1`] = `"param0, param1, param2, param3, param4, param5, param6, param7, param8, param9, param10..."`;
